solidity/contracts/fully-backed/FullyBackedBonding.sol

```diff
@@ -0,0 +1,191 @@
+/// of ETH the owner is willing to delegate should be deposited for the given
+/// operator.
+contract FullyBackedBonding is
+    IFullyBackedBonding,
+    AbstractBonding,
+    Authorizations,
+    StakeDelegatable
+{
+    event Delegated(address indexed owner, address indexed operator);
+
+    event OperatorDelegated(
+        address indexed operator,
+        address indexed beneficiary,
+        address indexed authorizer
+    );
+
+    // The ether value (in wei) that should be passed along with the delegation
+    // and deposited for bonding.
+    uint256 public constant MINIMUM_DELEGATION_DEPOSIT = 12345; // TODO: Decide right value
+
+    // Once a delegation to an operator is received the delegator has to wait for
+    // specific time period before being able to pull out the funds.
+    uint256 public constant DELEGATION_LOCK_PERIOD = 14 days; // TODO: Decide right value
+
+    uint256 public initializationPeriod; // varies between mainnet and testnet
+
+    /// @notice Initializes Fully Backed Bonding contract.
+    /// @param _keepRegistry Keep Registry contract address.
+    /// @param _initializationPeriod To avoid certain attacks on group selection,
+    /// recently delegated operators must wait for a specific period of time
+    /// before being eligible for group selection.
+    constructor(KeepRegistry _keepRegistry, uint256 _initializationPeriod)
+        public
+        AbstractBonding(address(_keepRegistry))
+        Authorizations(_keepRegistry)
+    {
+        initializationPeriod = _initializationPeriod;
+    }
+
+    /// @notice Registers delegation details. The function is used to register
+    /// addresses of operator, beneficiary and authorizer for a delegation from
+    /// the caller.
+    /// The function requires ETH to be submitted in the call as a protection
+    /// against attacks blocking operators. The value should be at least equal
+    /// to the minimum delegation deposit. Whole amount is deposited as operator's
+    /// unbonded value for the future bonding.
+    /// @param operator Address of the operator.
+    /// @param beneficiary Address of the beneficiary.
+    /// @param authorizer Address of the authorizer.
+    function delegate(
+        address operator,
+        address payable beneficiary,
+        address authorizer
+    ) external payable {
+        address owner = msg.sender;
+
+        require(
+            operators[operator].owner == address(0),
+            "Operator already in use"
+        );
+
+        require(
+            msg.value >= MINIMUM_DELEGATION_DEPOSIT,
+            "Insufficient delegation value"
+        );
+
+        operators[operator] = Operator(
+            OperatorParams.pack(0, block.timestamp, 0),
+            owner,
+            beneficiary,
+            authorizer
+        );
+
+        deposit(operator);
+
+        emit Delegated(owner, operator);
+        emit OperatorDelegated(operator, beneficiary, authorizer);
+    }
+
```

Comment:
We've got `deposit` function on bonding contract that emits `UnbondedValueDeposited`. I'm not sure about duplicating it. What if someone top-ups directly with `deposit` function? This is how we handle it currently for KEEP backed bonding.
To calculate the values we will need to analyze events, we started working on that in https://github.com/keep-network/local-setup/pull/48.

Reply:
I am also not sold on this idea but I also can't imagine analyzing all the events when rendering a page in token dashboard, it would take too much time.

Reply:
I can imagine that we will have to analyze  in the dashboard all the events anyway, to cover cases where top-up function was omitted and eth has been depositet directly.

Reply:
Yesterday, I tried to run the script from `local-setup` analyzing bonds per event locally against Infura. One hour was not enugh to produce results for ~100 mainnet operators.

Reply:
I'm not sure how much optimized is the script you mentioned. Tracking the value in contract instead of calculating it from events would be the fastest solution.

I believe we just have to agree to disagree. I'll add a top-up function you wanted.

Reply:
> Tracking the value in contract instead of calculating it from events would be the fastest solution.

We'd need a separate function to return bonds to the bonding contract. So `deposit` would be used only by the user and `returnBonds` (or any other function name) would be used to return partial bonds by application. Too late for this one and it still means we need two functions and there is no guarantee user calls the right one.